mesa/agent.py

```diff
@@ -57,3 +63,94 @@ def advance(self) -> None:
     @property
     def random(self) -> Random:
         return self.model.random
+
+
+class AgentSet(MutableSet):
+    """Ordered set of agents"""
+
+    def __init__(self, agents: Iterable[Agent], model: Model):
+        self._agents = weakref.WeakKeyDictionary()
+        for agent in agents:
+            self._agents[agent] = None
+
+        self.model = model
+
+    def __len__(self) -> int:
+        return len(self._agents)
+
+    def __iter__(self) -> Iterator[Agent]:
+        return self._agents.keys()
+
+    def __contains__(self, agent: Agent) -> bool:
+        """Check if an agent is in the AgentSet."""
+        return agent in self._agents
+
+    def select(self, filter_func: Callable[[Agent], bool] | None = None, inplace: bool = False) -> AgentSet:
+        if filter_func is None:
+            return self if inplace else AgentSet(list(self._agents.keys()), self.model)
+
+        agents = [agent for agent in self._agents.keys() if filter_func(agent)]
+
+        if inplace:
+            self._reorder(agents)
+            return self
+        else:
+            return AgentSet(
+                agents,
+                self.model
+            )
+
+    def shuffle(self, inplace: bool = False)-> AgentSet:
+        shuffled_agents = list(self._agents.keys())
+        self.model.random.shuffle(shuffled_agents)
+
+        if inplace:
+            self._reorder(shuffled_agents)
+            return self
+        else:
+            return AgentSet(shuffled_agents, self.model)
+
+    def sort(self, key: Callable[[Agent], Any], reverse: bool = False, inplace: bool = False)-> AgentSet:
+        sorted_agents = sorted(self._agents.keys(), key=key, reverse=reverse)
+
+        if inplace:
+            self._reorder(sorted_agents)
+            return self
+        else:
+            return AgentSet(sorted_agents, self.model)
+
+    def _reorder(self, agents: Iterable[Agent]):
+        _agents = weakref.WeakKeyDictionary()
+        for agent in agents:
+            _agents[agent] = None
+
+        self._agents = _agents
+
+    def do_each(self, method_name: str, *args, **kwargs) -> list[Any]:
+        """invoke method on each agent"""
+        return [getattr(agent, method_name)(*args, **kwargs) for agent in self._agents]
+
+    def get_each(self, attr_name: str) -> list[Any]:
+        """get attribute value on each agent"""
+        return [getattr(agent, attr_name) for agent in self._agents]
+
+    def __getitem__(self, item: int | slice) -> Agent:
+        return list(self._agents.keys())[item]
```

Comment:
Converting all agents to a list seems a bit inefficient for retrieving a single agent

```suggestion
        if isinstance(item, slice):
            return list(self._agents.keys())[item]
        return next(itertools.islice(self._agents.keys(), item, item+1))
```

Reply:
@quaquel what do you think?

Reply:
This concerns how we want to handle agents disappearing due to the garbage collector. 

I checked various other ordered set and indexed set implementations. They either do something like the current code or have a way of keeping track of changes to the internal dict, which then triggers a rebuilding of the index. I can do some performance tests on the suggested code and my version to see if it makes a big difference.

Reply:
Just to elaborate a bit on the problem.

Because of weakrefs, agents can magically disappear from `self._agent`. Normal OrderedSet implementations typically have a list of keys, and indexing is done on this. If items are added or removed from the set, the index is updated. Because of weakrefs, we don't fully know when agents are removed from the set. Thus we cannot do the default updating. 

One option is to replace the list typically used with a WeakValueDict, but you cannot conveniently slice on this. You would have to turn the slice into a list of indices and next for each index retrieve the key. Also, you can potentially retrieve None's if the index refers to an agent that has been garbage collected. So, how to handle such None's?

The other option, which I don't know will work, but I can check, is to use weakref. finalizers. In my understanding this allows us to detect when agents are removed by the garbage collector, in turn triggering a rebuilding of the index.

Whether any of this makes a big difference in performance, I also don't know yet.

Reply:
I see, its more complicated than I thought. I would say keep the current code, performance optimization (if needed!) can come later.

Reply:
Resolving for now